### ISSUE TITLE
[Epic] New Contract Dependency Manager

### ISSUE THREAD

**Issue by nvdtf** [onflow/flow-cli#929]
## Overview
Following the declarative design principles of Flow CLI we want to build a simple dependency manager that enables developers to:

- Import a known contract from supported locations such as Flow networks (mainnet, testnet, …), GitHub repositories, or the NFT Catalog into the current project.
- Deliver code with dependencies without re-shipping copies of dependencies.
- Checkout missing dependencies based on `flow.json`

## Design

Each contract is identified in `flow.json` by a unique identifier. It can be imported into another contract, transaction, or script like this:

```js
import "MyContract"
```

The pre-processor will replace the imports with valid Cadence syntax and addresses based on `flow.json`:

```js
"contracts": {
	"MyContract": "./contracts/MyContract.cdc", // local file location
	"TopShot": "mainnet/0x0123123123/TopShot", // live network address
	"NonFungibleToken": "github.com/onflow/flow-nft/NonFungibleToken.cdc", // github
	"CoolNFT": "nft-catalog/mainnet/CoolNFT"
}
```

The simple syntax can be evolved in the future to support specific commit hashes or versions for GitHub, or block heights for addresses.
Flow CLI will support checking for missing dependencies and checking them out into a local `imports` folder. The folder can be added to `.gitignore`so other developers always pull the dependencies from the original source-of-truth.

### `flow contracts install`

Flow CLI can check all dependencies and fetch missing ones into `imports` with this command:

```bash
flow contracts install
```

### `flow contracts add`

- From live Flow networks:

```bash
flow contracts add mainnet/0x0123123123/TopShot
```

A copy of the contract is fetched into the `imports` folder. For each import, a subfolder is created and the same operation is executed recursively. An `alias` entry is also created in `flow.json` to prevent re-deployments. Check out 
[cadence-import](https://github.com/jrkhan/cadence-import) to see this in action.

- From NFT Catalog:

```bash
flow contracts add nft-catalog/mainnet/CoolNFT
```

The contract address is resolved via NFT Catalog. The same steps as importing from networks are applied.

- From GitHub:

```bash
flow contracts add github.com/onflow/flow-nft/NonFungibleToken.cdc
```

A copy of the contract is fetched into the `imports` folder from the latest commit. The name of the contract will be resolved from the contents. If the contract has imports, the repository’s root `flow.json` is checked to resolve them.

```[tasklist]
### Issues
- [ ] #910 
- [ ] #331 
- [ ] #944
- [ ] #957
- [ ] #958 
```

# References

- https://github.com/jrkhan/cadence-import

**Comment by bjartek** [comment]
Note that some contracts are very unfriendly to deploy and install on emulator, if they have very many arguments to the initializer. FiatToken i am looking at you. 

We have solved this now by simple providing a shim for it that is a lot simpler. I would very much like the system to support  shims for emulators that behave as the deployed version but it a lot easier to work with.

**Comment by btspoony** [comment]
For GitHub import, can we introduce the concept of exporting contracts? 
There will be no ambiguity when importing.

Firstly, add an 'export' field in flow.json.
```jsonc
{
  "contracts": {
    "MyContract": "./contracts/PathToFolder/MyContract.cdc", // local file location, 
  },
  "exports": [
    "MyContract"
  ]
}
```

Then when using it, do as follows:
```jsonc
{
  "contracts": {
    "MyContract": "github.com/username/reponame/MyContract", // GitHub, can only import contracts from exports, no worry about the path
  },
}
```

**Comment by bluesign** [comment]
I think we need to use more advanced format, something like: 

```json
{
    "contracts": {
        "MyContract": {
            "type" : "local",
            "localPath": "./contracts/MyContract.cdc"
        },
        "TopShot":{
            "type": "flow",
            "localPath": "./contracts/utility/TopShot.cdc",
            "remotePath":{
                "mainnet": "0x0123123123/TopShot",
                "testnet": "0x0424242424/TopShot"
            }
        },
        "NonFungibleToken":{
            "type": "github",
            "localPath": "./contracts/utility/NonFungibleToken.cdc",
            "remotePath":{
                "mainnet": "github.com/onflow/flow-nft/NonFungibleToken.cdc",
                "testnet": "github.com/onflow/flow-nft/NonFungibleToken.cdc"
            },
            "aliases":{
                "mainnet": "0x0123123123",
                "testnet": "0x0424242424"
            }
        },
        "CoolNFT":{
            "type": "nftcatalog",
            "localPath": "./contracts/CoolNFT.cdc",
            "remotePath":{
                "mainnet": "mainnet/CoolNFT",
                "testnet": "testnet/CoolNFT"
            },
            "aliases":{
                "mainnet": "0x0123123123",
                "testnet": "0x0424242424"
            }
        },
    }
}
```

**Comment by nvdtf** [comment]
> Note that some contracts are very unfriendly to deploy and install on emulator, if they have very many arguments to the initializer. FiatToken i am looking at you.
> 
> We have solved this now by simple providing a shim for it that is a lot simpler. I would very much like the system to support shims for emulators that behave as the deployed version but it a lot easier to work with.

@bjartek You're right. I created another issue for this, since it seems like there is no support for contract init params: https://github.com/onflow/flow-cli/issues/964
Feel free to comment if you have any suggestions on how to change `flow.json` for this.

> Thanks @nvdtf this would be quite helpful for local development, and might also help improve contract deployment pipelines. Would also like to suggest `-n` should be a valid alternative for specifying network in the live flow Network flavor as it would align with usage for other commands:
> 
> ```
> flow contracts add 0x0123123123/TopShot -n mainnet
> ```

@jrkhan I re-used the simple syntax from `flow.json` for simplicity. We can also check if `-n` is present and support dropping the network name from the contract path. This can work with NFT catalog entries too. Thanks for the suggestion!

> For GitHub import, can we introduce the concept of exporting contracts? There will be no ambiguity when importing.
> 
> Firstly, add an 'export' field in flow.json.
> 
> ```js
> {
>   "contracts": {
>     "MyContract": "./contracts/PathToFolder/MyContract.cdc", // local file location, 
>   },
>   "exports": [
>     "MyContract"
>   ]
> }
> ```
> 
> Then when using it, do as follows:
> 
> ```js
> {
>   "contracts": {
>     "MyContract": "github.com/username/reponame/MyContract", // GitHub, can only import contracts from exports, no worry about the path
>   },
> }
> ```

@btspoony The problem with supporting exports is that the original developer must be mindful of future reusability. I think we need to have `exports`, but with higher-level constructs like JS packages or FLIX interactions. I don't see a problem with being able to import any given contract from anywhere.

> I think we need to use more advanced format, something like:
> 
> ```json
> {
>     "contracts": {
>         "MyContract": {
>             "type" : "local",
>             "localPath": "./contracts/MyContract.cdc"
>         },
>         "TopShot":{
>             "type": "flow",
>             "localPath": "./contracts/utility/TopShot.cdc",
>             "remotePath":{
>                 "mainnet": "0123123123.TopShot",
>                 "testnet":  "042424242.TopShot"
>             }
>         },
>         "NonFungibleToken":{
>             "type": "github",
>             "localPath": "./contracts/utility/NonFungibleToken.cdc",
>             "remotePath":{
>                 "mainnet": "github.com/onflow/flow-nft/NonFungibleToken.cdc",
>                 "testnet": "github.com/onflow/flow-nft/NonFungibleToken.cdc"
>             },
>             "aliases":{
>                 "mainnet": "0x0123123123",
>                 "testnet": "0x0424242424"
>             }
>         },
>         "CoolNFT":{
>             "type": "nftcatalog",
>             "localPath": "./contracts/CoolNFT.cdc",
>             "remotePath":{
>                 "mainnet": "mainnet/CoolNFT",
>                 "testnet": "testnet/CoolNFT"
>             },
>             "aliases":{
>                 "mainnet": "0x0123123123",
>                 "testnet": "0x0424242424"
>             }
>         },
>     }
> }
> ```
> 
> also deploy shim idea is also great ( or some init args in flow.json )

@bluesign Definitely looking to expand the simple syntax into advanced with more config options. I was also thinking of branch commit hash, content hash, ...
I'll open another issue in the future to expand the syntax.

**Comment by btspoony** [comment]
New thoughts:
https://discord.com/channels/613813861610684416/1123314820763111465/1164943141699342527

**Comment by chasefleming** [comment]
The initial proposal is clean, but I think there is a big gap from going from something like that to the extended format (which seems required to actually have enough information across networks to work). Also, in a more extended format I think we can lean in more to the existing terminology and structure a developer already knows as to not overwhelm in difficulty and confusion. We could do something like this:

```
{
  "contracts": {
    // local contracts
  },
  "dependencies": {
    "TopShot": {
      "path": "./imports/TopShot",
      "remotePath": "testnet/0x0123123123.TopShot",
      "aliases": {
        "testnet": "0x0123123123",
        "mainnet": "0x0123123124"
      }
    }
  },
  "networks": {
    "emulator": "127.0.0.1:3569",
    "mainnet": "access.mainnet.nodes.onflow.org:9000",
    "testnet": "access.devnet.nodes.onflow.org:9000"
  },
  "accounts": {
    // account configurations
  }
}
```

In this the dependencies from the contract manager that you are not developing locally are broken out into a separate `dependencies` section so it's clear there is a difference and that different key/values may be possible. Then since there is an issue with different networks having different contract versions we should let the developer define the single source of truth they want to use to pull and develop with locally, which is the `removePath`. Aliases are then also still defined for running against `testnet`, `mainnet`, or any other network. The nice part is that this is already the pattern they are familiar with from the `contracts` section.

A developer should not set `path`, so the initial adding of a contract would look just like this:

```
"dependencies": {
    "TopShot": {
      "remotePath": "testnet/0x0123123123.TopShot",
    }
  },
  ```

Then someone could run `flow contracts install` and the `path` would be created. Curious to hear feedback.

**Comment by nvdtf** [comment]
@chasefleming  We can remove `path` in `dependencies` since the local path will be dictated automatically (like `dependencies/mainnet/0x1/A.cdc`). We also don't expect the developer to set a custom path.
For simplicity, we can also always check out the mainnet version of a dependency locally when it's defined on both networks.

**Comment by chasefleming** [comment]
Yeah I agree `path` seems unnecessary and we don't want developer manually editing them by mistake. Let's remove that. Based on the feedback, it would be this:

```
"dependencies": {
		"TopShot": {
			"remoteSource": "testnet/0x0123123123.TopShot",
			"aliases": {
				"testnet": "0x0123123123",
				"mainnet": "0x0123123124"
			}
		}
	}
```

@bjartek I also agree a mapping would be nice, but I'm not sure we can rely on every contract having it so we'd need to have this working without that first and if there was a way to group down the road that could speed things up for a developer.

**Comment by bjartek** [comment]
If you cant trust the author of a contract you should not use it. You have to trust somebody. And since this just populates local files and flow.json you can always validate right?

**Comment by bjartek** [comment]
If you cant trust the author of a contract you should not use it. You have to trust somebody. And since this just populates local files and flow.json you can always validate right?

**Comment by bjartek** [comment]
Another option besides pragma is a resource you can store in your accoung. We talked about this in the flix wg yesterday, but not sure how well it would scale. @JeffreyDoyle

**Comment by nvdtf** [comment]
I was thinking we only define dependencies so:

- Their addresses are aliases and translated correctly on each network since we won't be deploying them again alongside our own contracts.
- We can run them in the emulator locally.

On the first point, if a given contract's mainnet/testnet code differs, we aren't impacted since we rely only on translated import addresses. The checked-out code is only useful for the second point, local deployment. This means that we're grabbing contracts from a source of truth (mainnet/testnet) and will run it in the emulator. We can argue at this point if we have a contract that has a different code on mainnet/testnet it won't probably work in emulator anyway since it's relying on network addresses in its code. Based on this point, we can have a rule like "always get the mainnet code" and simplify. Let me know what you think.

**Comment by bjartek** [comment]
Some projects that you might want to implement might not be on mainnet yet. We need to cover that case to. 

What to do with Fiat token for instance? Personally i think USDC should bust be on emulator like FlowToken.

**Comment by nvdtf** [comment]
We need to support contract init args before we think about `FiatToken` 🥲
I agree it should be emulator. I support adding lots of more base contracts to the emulator in favor of better DX.

**Comment by bjartek** [comment]
Personally i think flow init should have a flag to populate flow.json with all the aliases that are deployed when you start emulator. That makes bootstrapping way easier.

**Comment by bjartek** [comment]
But we are not now only talking about a dependenxcy manager for contracts. What about flix/tx/scripts to initialize an «addon».

**Comment by chasefleming** [comment]
As I began coding and reached the step for deploying dependencies, I realized that instead of trying to integrate dependencies into commands like `flow project deploy`, it's more seamless to build upon the existing `flow.json` patterns. In these patterns, you list dependencies that ultimately get added to contracts. This approach ensures that everything continues to work smoothly with the Flow CLI. I've modified the code I'm working on to operate as follows:

Initially, you would list dependencies in this format, which is easier to add (I will incorporate the `://` suggestion from @bjartek mentioned during office hours):

```
{
	"contracts": {
		"HelloWorld": "cadence/contracts/HelloWorld.cdc"
	},
	"dependencies": {
		"TopShot": "testnet://877931736ee77cff.TopShot"
	},
	"networks": {
		"emulator": "127.0.0.1:3569",
		"mainnet": "access.mainnet.nodes.onflow.org:9000",
		"testnet": "access.devnet.nodes.onflow.org:9000"
	},
	"accounts": {
		"emulator-account": {
			"address": "f8d6e0586b0a20c7",
			"key": "1a42a3e48bd7f09ae61e9d39e2cf7045f5499424468dcf8d7e39f62426e238d0"
		}
	}
}
```

After running `install`, the result will be as follows, where all dependencies are listed, and the contracts section is populated with your dependencies:

```
{
	"contracts": {
		"FungibleToken": {
			"source": "imports/9a0766d93b6608b7/FungibleToken",
			"aliases": {
				"testnet": "9a0766d93b6608b7"
			}
		},
		"HelloWorld": "cadence/contracts/HelloWorld.cdc",
		"MetadataViews": {
			"source": "imports/631e88ae7f1d7c20/MetadataViews",
			"aliases": {
				"testnet": "631e88ae7f1d7c20"
			}
		},
		"NonFungibleToken": {
			"source": "imports/631e88ae7f1d7c20/NonFungibleToken",
			"aliases": {
				"testnet": "631e88ae7f1d7c20"
			}
		},
		"TopShot": {
			"source": "imports/877931736ee77cff/TopShot",
			"aliases": {
				"testnet": "877931736ee77cff"
			}
		},
		"TopShotLocking": {
			"source": "imports/877931736ee77cff/TopShotLocking",
			"aliases": {
				"testnet": "877931736ee77cff"
			}
		}
	},
	"dependencies": {
		"FungibleToken": "testnet//:9a0766d93b6608b7.FungibleToken",
		"MetadataViews": "testnet//:631e88ae7f1d7c20.MetadataViews",
		"NonFungibleToken": "testnet//:631e88ae7f1d7c20.NonFungibleToken",
		"TopShot": "testnet//:877931736ee77cff.TopShot",
		"TopShotLocking": "testnet//:877931736ee77cff.TopShotLocking"
	},
	"networks": {
		"emulator": "127.0.0.1:3569",
		"mainnet": "access.mainnet.nodes.onflow.org:9000",
		"testnet": "access.devnet.nodes.onflow.org:9000"
	},
	"accounts": {
		"emulator-account": {
			"address": "f8d6e0586b0a20c7",
			"key": "1a42a3e48bd7f09ae61e9d39e2cf7045f5499424468dcf8d7e39f62426e238d0"
		}
	}
}
```

Now, you can run `flow project deploy` without encountering any issues.

**Comment by bjartek** [comment]
This makes sense to me, have you considered using either identifiers for imports or replace the . With /? So either

testnet://A.9a0766d93b6608b7.FungibleToken

Or
testnet://0x9a0766d93b6608b7/FungibleToken
